src/D2L.CodeStyle.Analyzers/ApiUsage/Events/EventHandlersBlacklistAnalyzer.cs

```diff
@@ -0,0 +1,125 @@
+using System.Collections.Immutable;
+using System.Linq;
+using D2L.CodeStyle.Analyzers.Extensions;
+using Microsoft.CodeAnalysis;
+using Microsoft.CodeAnalysis.CSharp;
+using Microsoft.CodeAnalysis.CSharp.Syntax;
+using Microsoft.CodeAnalysis.Diagnostics;
+
+namespace D2L.CodeStyle.Analyzers.ApiUsage.Events {
+
+	[DiagnosticAnalyzer( LanguageNames.CSharp )]
+	internal sealed class EventHandlersBlacklistAnalyzer : DiagnosticAnalyzer {
+
+		public override ImmutableArray<DiagnosticDescriptor> SupportedDiagnostics => ImmutableArray.Create(
+				Diagnostics.EventHandlerBlacklisted
+			);
+
+		public override void Initialize( AnalysisContext context ) {
+
+			context.EnableConcurrentExecution();
+			context.RegisterCompilationStartAction( RegisterAnalysis );
+		}
+
+		private void RegisterAnalysis( CompilationStartAnalysisContext context ) {
+
+			Compilation compilation = context.Compilation;
+
+			IImmutableSet<INamedTypeSymbol> blacklistedTypes = GetBlacklistedTypes( compilation );
+			if( blacklistedTypes.Count == 0 ) {
+				return;
+			}
+
+			context.RegisterSyntaxNodeAction(
+					c => AnalyzeSimpleBaseType( c, blacklistedTypes ),
+					SyntaxKind.SimpleBaseType
+				);
+		}
+
+		private void AnalyzeSimpleBaseType(
+				SyntaxNodeAnalysisContext context,
+				IImmutableSet<INamedTypeSymbol> blacklistedTypes
+			) {
+
+			SimpleBaseTypeSyntax baseTypeSyntax = (SimpleBaseTypeSyntax)context.Node;
+			SymbolInfo baseTypeSymbol = context.SemanticModel.GetSymbolInfo( baseTypeSyntax.Type );
+
+			INamedTypeSymbol baseSymbol = ( baseTypeSymbol.Symbol as INamedTypeSymbol );
+			if( baseSymbol.IsNullOrErrorType() ) {
+				return;
+			}
+
+			if( !blacklistedTypes.Contains( baseSymbol ) ) {
+				return;
+			}
+
+			Diagnostic diagnostic = Diagnostic.Create(
+					Diagnostics.EventHandlerBlacklisted,
+					baseTypeSyntax.GetLocation(),
+					baseSymbol.ToDisplayString()
+				);
+
+			context.ReportDiagnostic( diagnostic );
+		}
+
+		private static IImmutableSet<INamedTypeSymbol> GetBlacklistedTypes( Compilation compilation ) {
+
+			IImmutableSet<INamedTypeSymbol> types = EventHandlersBlacklist.BlacklistedTypes
+				.SelectMany( genericType => GetGenericTypes( compilation, genericType.Key, genericType.Value ) )
+				.ToImmutableHashSet();
+
+			return types;
+		}
+
+		private static IEnumerable<INamedTypeSymbol> GetGenericTypes(
+				Compilation compilation,
+				string genericTypeName,
+				ImmutableArray<ImmutableArray<string>> genericTypeArgumentSets
+			) {
+
+			INamedTypeSymbol genericTypeDefinition = compilation.GetTypeByMetadataName( genericTypeName );
+			if( genericTypeDefinition.IsNullOrErrorType() ) {
+				yield break;
+			}
+
+			foreach( ImmutableArray<string> genericTypeArguments in genericTypeArgumentSets ) {
+				if( TryGetGenericType( compilation, genericTypeDefinition, genericTypeArguments, out INamedTypeSymbol genericType ) ) {
+					yield return genericType;
+				}
+			}
+		}
+
+		private static bool TryGetGenericType(

```

Comment:
Ok I see why you had `ImmutableArray<ImmutableArray<string>>` up there, multiple generic type arguments (but we only ever have one for the two event handlers.)

Don't you think it'd be simpler to just have

```cs
ImmutableHashSet<INamedTypeSymbol> badThings = ImmutableArray.Create(
  "D2L.LP.Distributed.Events.Handlers.IEventHandler<D2L.Core.JobManagement.Events.JobAbortRequested>",
  "D2L.LP.Distributed.Events.Handlers.IOrgEventHandler<D2L.Core.JobManagement.Events.JobAbortRequested>",

  "D2L.LP.Distributed.Events.Handlers.IEventHandler<D2L.LP.Distributed.WorkQueues.Events.WorkQueueEvent>",
  "D2L.LP.Distributed.Events.Handlers.IOrgEventHandler<D2L.LP.Distributed.WorkQueues.Events.WorkQueueEvent>",

  // ...
  )
  .Select( t => compilation.GetTypeByMetadataName )
  .ToImmutableHashSet();
```

with none of the (sorta complicated) helper methods. Obviously there's some duplication in there (the generic type name) and you have to be "careful" to add both but it's just so much less code. That's what I was trying to get at in the previous comment.

If you're attached to the way it's working now that's fine, and you can go back to what you had before if you want. The difference between this and what I remember in the initial PR is pretty small and doesn't really hit what I was talking about, so either way IMO.

Reply:
I had originally started with the duplication, but decided it was better to have a little more code and remove the human error prone aspect.